src/oom/oomd-util.c

```diff
@@ -282,6 +283,204 @@ int oomd_cgroup_kill(const char *path, bool recurse, bool dry_run) {
         return !set_isempty(pids_killed);
 }
 
+static void oom_kill_state_free(OomdKillState *ks) {
+        sd_varlink *link;
+
+        assert(ks);
+        assert(ks->manager);
+
+        SET_FOREACH(link, ks->links)
+                sd_varlink_unref(link);
+        set_free(ks->links);
+
+        set_remove(ks->manager->prekill_ctxs, ks);
+        oomd_cgroup_context_unref(ks->ctx);
+        free(ks);
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(OomdKillState*, oom_kill_state_free, NULL);
+
+static OomdKillState* oomd_kill_state_dec(OomdKillState *ks) {
+        int r;
+
+        assert(ks);
+        assert(ks->ctx->path);
+
+        r = oomd_cgroup_kill(ks->ctx->path, /* recurse= */ true);
+        if (r < 0)
+                log_debug_errno(r, "Failed to kill cgroup '%s', ignoring: %m", ks->ctx->path);
+        oom_kill_state_free(ks);
+        return NULL;
+}
+
+DEFINE_PRIVATE_TRIVIAL_REF_UNREF_FUNC(OomdKillState, oomd_kill_state, oomd_kill_state_dec);
+
+int clean_prekills(struct Set *prekill_ctxs) {
+        OomdKillState *ks;
+
+        SET_FOREACH(ks, prekill_ctxs) {
+                log_debug("Cleaning up unfinished prekill hook state for cgroup %s", ks->ctx->path);
+                oom_kill_state_free(ks);
+        }
+
+        return 0;
+}
+
+static int prekill_callback(
+                sd_varlink *link,
+                sd_json_variant *parameters,
+                const char *error_id,
+                sd_varlink_reply_flags_t flags,
+                void *userdata) {
+
+        OomdKillState *ks = ASSERT_PTR(userdata);
+
+        assert(ks->ctx->path);
+
+        if (error_id)
+                log_warning("oomd prekill hook for %s returned error: %s", ks->ctx->path, error_id);
+        else
+                log_info("oomd prekill hook finished for cgroup %s", ks->ctx->path);
+
+        set_remove(ks->links, link);
+        oomd_kill_state_unref(ks);
+        sd_varlink_unref(link);
+
+        return 0;
+}
+
+static int send_prekill_message(
+                const char *basename,
+                sd_json_variant *cparams,
+                OomdKillState *ks,
+                sd_event *e) {
+
+        _cleanup_(sd_varlink_close_unrefp) sd_varlink *link = NULL;
+        _cleanup_free_ char *hook_path = NULL;
+        int r;
+
+        assert(basename);
+        assert(cparams);
+        assert(e);
+        assert(ks);
+        assert(ks->manager);
+
+        log_info("Invoking oomd prekill hook %s for cgroup %s", basename, ks->ctx->path);
+
+        hook_path = path_join(VARLINK_ADDR_PATH_OOMD_PREKILL, basename);
+        if (!hook_path)
+                return log_oom_debug();
+
+        r = sd_varlink_connect_address(&link, hook_path);
+        if (r < 0)
+                return log_debug_errno(r, "Prekill socket at %s not connected, ignoring.", hook_path);
```

Comment:
You're not ignoring the error here. You're actually returning a error. This message is misleading.

IMO, you should `return log_debug_errno()` everywhere in this function. And then `log_warning_errno(r, "Failed to call prekill hook %s, ignoring")` where you call `send_prekill_message()`